LibWeb: Implement `<feComposite>` SVG filter

tcl3 opened the pull request:
I have added a FIXME for the arithmetic operator, which I couldn't get working :sweat_smile: 

Fixes #5793

tcl3 committed:
LibWeb: Implement `<feComposite>` SVG filter

tcl3 commented:
My latest push adds the arithmetic filter, but I'm still getting a some color differences compared to other browsers.

I've tested the arithmetic filter with the following SVG:
```html
<svg width="100" height="100" fill="white" xmlns="http://www.w3.org/2000/svg">
<defs>
  <filter id="compositeFilter">
    <feFlood flood-color="white" result="whiteFill" />
    <feComposite in2="whiteFill" in="SourceGraphic" operator="arithmetic" k1=".1" k2=".2" k3=".3" k4=".4" />
  </filter>
</defs>

<rect x="0" y="0" width="100" height="100" fill="#000000" filter="url(#compositeFilter)" />
</svg>
```

If we treat the image as monochrome for this example the values for the relevant variables are:
* in1 = 0 (black)
* in2 = 1 (white)
* k1 = 0.1
* k2 = 0.2
* k3 = 0.3
* k4 = 0.4

Given the equation for the arithmetic filter:
 `(k1 * in1 * in2) + (k2 * in1) + (k3 * in2) + k4`
 
Python tells me the value I should get out is 0.7.

According to the GIMP eyedropper tool the value I get for each color channel is .698 - I assume the difference is due to quantization. However, the value that all other browsers get according to the eyedropper tool is 0.855.

I'm not sure why other browsers produce this brighter value.

gmta commented:
Results from Chrome on macOS:

<details><summary>Testcase SVG</summary>

NOTE: I've swapped in/in2 compared to the testcase mentioned earlier.

```
<svg width="100" height="100" fill="white" xmlns="http://www.w3.org/2000/svg">
<defs>
  <filter id="compositeFilter">
    <feFlood flood-color="white" />
    <feComposite in2="SourceGraphic" operator="arithmetic" k1="0" k2="0" k3="0.3" k4="0.4" />
  </filter>
</defs>

<rect x="0" y="0" width="100" height="100" fill="#000000" filter="url(#compositeFilter)" />
</svg>
```

</details>

| k4=0.4, k3: | 0.0 | 0.1 | 0.2 | 0.3 | 0.4 | 0.5 | 0.6 | 0.7 | 0.8 | 0.9 | 1.0 |
|--|--|--|--|--|--|--|--|--|--|--|--|
| Value | 1.0 | 0.953 | 0.902 | 0.847 | 0.788 | 0.729 | 0.667 | 0.667 | 0.667 | 0.667 | 0.667 |

| k3=0.3, k4: | 0.0 | 0.1 | 0.2 | 0.3 | 0.4 | 0.5 | 0.6 | 0.7 | 0.8 | 0.9 | 1.0 |
|--|--|--|--|--|--|--|--|--|--|--|--|
| Value | 0.698 | 0.816 | 0.831 | 0.843 | 0.847 | 0.851 | 0.851 | 0.855 | 0.906 | 0.957 | 1.0 |

Looking at these values, sure looks like some curve is being applied. Gamma application or sRGB -> linear conversion?

gmta commented:
Based on the partial calculation `(k3 * in2) + k4`, if I apply the sRGB-linear -> sRGB transformation I get values that seem to correspond a bit better:

`sRGB = 1.055 * pow(sRGB_linear, 1 / 2.4) - 0.055;`

This gives me `0.853` for `k3=0.3,k4=0.4` and `0.665` for `k3=1.0,k4=0.4`.

tcl3 commented:
I think you're right, the difference appears to be a linear RGB -> sRGB conversion from the application of the [`color-interpolation-filters` property](https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Attribute/color-interpolation-filters), which has a default value of `linearRGB`. 

If I set ` <filter color-interpolation-filters="sRGB ...>`, then we get the same value as Chrome/Firefox.

konradekk commented on Libraries/LibWeb/SVG/SVGFECompositeElement.idl:
```diff
@@ -0,0 +1,26 @@
+
+    readonly attribute SVGAnimatedString in1;
+    readonly attribute SVGAnimatedString in2;
+    [ImplementedAs=operator_for_bindings] readonly attribute SVGAnimatedEnumeration operator_for_bindings;
```
wasn't that
```suggestion
    [ImplementedAs=operator_for_bindings] readonly attribute SVGAnimatedEnumeration operator;
```

konradekk reviewed (commented):
the filter itself looks legit; the color space issues could be taken care of separately? (a todo somewhere would be enough?)

tcl3 commented:
> the color space issues could be taken care of separately? (a todo somewhere would be enough?)

I have an implementation of `color-interpolation-filters`, which corrects the color difference, but it relies on changes in #5869 and #5856, so I'd prefer to add that as a separate PR.

tcl3 committed:
LibWeb: Convert blend test to use rectangle rather than circle

Using a circle isn't relevant to the outcome of the test and introduces
anti-aliasing artifacts.
